### ISSUE TITLE
DBRX model passes probabilities and not logits to the load balancer

### ISSUE THREAD

**Issue by rudolfwilliam** [huggingface/transformers#39055]
### System Info

The problem has nothing to do with my system.

### Who can help?

@ArthurZucker

### Information

- [x] The official example scripts
- [ ] My own modified scripts

### Tasks

- [x] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

Look into the code and verify that it is not compliant with the [intended load balancing](https://arxiv.org/pdf/2101.03961):

The signature of the load balancer in [modeling_dbrx](https://github.com/huggingface/transformers/blob/main/src/transformers/models/dbrx/modeling_dbrx.py) looks as follows:

```python
def load_balancing_loss_func(
    gate_logits: torch.Tensor,
    num_experts: int,
    top_k: int,
    attention_mask: Optional[torch.Tensor],
) -> torch.Tensor:
```

And also, the `gate_logits` are treated as such, meaning that the softmax is computed later in the code. However, the values (called `weights`) that are passed into that function are the expert probabilities and not the logits:

```python
weights = self.layer(hidden_states).softmax(dim=-1, dtype=torch.float32)
```

Thus, the load balancer takes the softmax of the logits twice, which is incorrect.

### Expected behavior

`gate_logits` should be logits and not probabilities.

**Comment by Rocketknight1** [comment]
This code hasn't been changed since the model was added in #29921 by @abhi-mosaic and @eitanturok. If either of those see this ping, can you verify? Especially @eitanturok since I believe he was/is working at Databricks.
